[AArch64][GlobalISel] Added support for hadd family of intrinsics

JoshdRod opened the pull request:
GlobalISel now selects hadd family of intrinsics, without falling back to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Added uhadd intrinsic support

GlobalISel now selects uhadd intrinsic, without falling back to SDAG.
Note that GlobalISel-generated code involving uhadd seems to be inefficent when compared to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Added urhadd intrinsic support

GlobalISel now selects urhadd intrinsic, without falling back to SDAG.
Note that GlobalISel-generated code involving urhadd seems to be inefficent when compared to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Added shadd intrinsic support
GlobalISel now selects shadd intrinsic, without falling back to SDAG.
Note that GlobalISel-generated code involving shadd seems to be inefficent when compared to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Added srhadd intrinsic support
GlobalISel now selects srhadd intrinsic, without falling back to SDAG.
Note that GlobalISel-generated code involving uhadd seems to be inefficent when compared to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Modified llc test to check generation from both SDAG and GISel
Note that GlobalISel-generated code involving the hadd family of intrinsics seems to be inefficent when compared to SDAG.

davemgreen reviewed (commented):
I think we should probably reuse the same names as is used in SDAG. As much as I dislike them, there are other "hadd" instructions on other architectures that do something a little different IIRC.

JoshdRod committed:
[AArch64][GlobalISel] Modified gMIR instruction names to match SDAG equivalents.

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrGISel.td:
```diff
@@ -239,6 +239,30 @@ def G_USDOT : AArch64GenericInstruction {
   let hasSideEffects = 0;
 }
 
+def G_AVGFLOORU : AArch64GenericInstruction {
```
Can we make these generic - add these to the same places as G_ABDU. We will need them eventually when we add combines for them in the same way as SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Converted intrinsics to machine-independent form

JoshdRod committed:
[AArch64][GlobalISel] Fixed formatting

JoshdRod committed:
[AArch64][GlobalISel] Renamed GISel nodes for consistency

davemgreen reviewed (commented):
Thanks, can you add something to llvm/docs/GlobalISel/GenericOpcode.rst, and can we add test support for llvm/test/CodeGen/AArch64/trunc-avg-fold.ll and llvm/test/CodeGen/AArch64/arm64-vhadd.ll too?

Otherwise LGTM

JoshdRod committed:
[GlobalISel] Added documentation for gMIR instructions into GenericOpcode.rst

davemgreen commented on llvm/docs/GlobalISel/GenericOpcode.rst:
```diff
@@ -511,6 +511,19 @@ Compute the absolute difference (signed and unsigned), e.g. trunc(abs(ext(x)-ext
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Computes the average of corresponding elements in two vectors (signed and unsigned).
+Resulting vector contains values that are either rounded or truncated. e.g. trunc(shr(ext(a)+ext(b))).
```
trunc(shr(add(ext(a),ext(b)),1))

JoshdRod replied:
Any reason why the example on ln 507 uses - instead of sub()?

JoshdRod committed:
[AArch64][GlobalISel] Modified trunc-avg-fold.ll to separately test SDAG and GISel generated code

Test file contains only CHECK-SD and CHECK-GI prefixes, as shared CHECK prefix is not needed.

JoshdRod committed:
[AArch64][GlobalISel] Modified arm64-vhadd.ll to separately test SDAG and GISel generated code

davemgreen reviewed (commented):
Thanks - I think the tests are failing on x86? There are some tests that only run on x86 that need updating, IIRC.

JoshdRod committed:
[AArch64][GlobalISel] Removed redundant AArch64-specific GI node definitions

Now that avg(floor/ceil)(u/s) are generic intrinsics, with GI node definitions in the generic SelectionDAGCompat.td, the AArch64-specific defintions have been removed.

JoshdRod committed:
[AArch64][GlobalISel] Fixed typo in Generic Opcode example

JoshdRod committed:
[RISCV][GlobalISel] Added intrinsics to legalizer-info-validation.mir

Test has been modified to include the new generic hadd intrinsics. hadd intrinsics do not lower on RISC-V.
Also fixed formatting on above lines in test.

JoshdRod committed:
[X86][GlobalISel] Patched vocabs-basic.ll test

First of 3 tests that need changing due to implementation of new generic intrinsics.

JoshdRod committed:
[X86][GlobalISel] Patched entities.mir test

Second of 3 tests that need chaning due to implementation of new generic intrinsics

JoshdRod committed:
[X86][GlobalISel] Patched triplets.mir test

Third of 3 tests that need changing due to implementation of new generic intrinsics

JoshdRod committed:
[AArch64][GlobalISel] Fixed failing test

JoshdRod committed:
[X86][GlobalISel] Updated entities.mir and triplets.mir to reflect upstream changes

davemgreen reviewed (approved):
Thanks. LGTM
